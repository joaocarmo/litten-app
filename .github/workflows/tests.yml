# This workflow will do a clean install of node dependencies, build the source
# code and run the available tests
name: Tests

on:
  push:
    branches: [ main, next ]
  pull_request:
    branches: [ main, next ]

env:
  CI: true
  FIREBASE_EMULATORS_PATH: $GITHUB_WORKSPACE/emulator-cache

jobs:
  test:
    environment: development
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v3
    - name: Use Node.js v16 (LTS)
      uses: actions/setup-node@v3
      with:
        node-version: '16'
        cache: 'yarn'
    - name: Install firebase-tools globally
      run: yarn global add firebase-tools
    - name: Create .env file
      run: cp .env.example .env
    - name: Install dependencies
      run: yarn --frozen-lockfile
    - name: Lint the source files
      run: yarn lint
    - name: Type check
      run: yarn type-check
    - name: Validate the translations
      run: yarn translations-check
    - name: Execute the tests
      run: yarn test
    - name: Execute the firebase rules tests
      run: yarn test:firestore:rules
    - name: Execute the firebase model tests
      run: yarn test:firestore:model
    - name: Cache firebase emulators
      uses: actions/cache@v3
      with:
        path: ${{ env.FIREBASE_EMULATORS_PATH }}
        key: ${{ runner.os }}-firebase-emulators-${{ hashFiles('emulator-cache/*.{jar,zip}') }}
      continue-on-error: true
