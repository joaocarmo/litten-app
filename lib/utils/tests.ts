import { apps, clearFirestoreData } from '@firebase/rules-unit-testing'

export const projectId = 'litten-app'

export const cleanFirestore = async (): Promise<void> => {
  await Promise.all(apps().map((app) => app.delete()))
  await clearFirestoreData({
    projectId,
  })
}

export const mapName =
  <T>(key: string) =>
  (object: T): T =>
    Object.assign(object, {
      toString: function () {
        return this[key]
      },
    })

export const getRandInt = (min: num, max: num): num =>
  Math.floor(Math.random() * (max - min + 1) + min)
