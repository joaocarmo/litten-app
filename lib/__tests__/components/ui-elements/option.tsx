import { render } from '@utils/tests/react-native'
import UISelectPlatform from '@ui-elements/inner-components/select'
import UISelectPlatformAndroid from '@ui-elements/inner-components/select/select.android'
import UISelectPlatformIOS from '@ui-elements/inner-components/select/select.ios'
import { UIOption } from '@ui-elements'

jest.mock('@ui-elements/inner-components/select')

const mockUISelectPlatform = UISelectPlatform as jest.MockedFunction<
  typeof UISelectPlatform
>

describe('Snapshot test for the "UISwitch" component', () => {
  const items = [
    {
      key: 'one',
      label: 'One',
      value: 'one',
    },
    {
      key: 'two',
      label: 'Two',
      value: 'two',
    },
  ]
  const selectedValue = 'one'

  const platform = [
    ['ios', UISelectPlatformIOS],
    ['android', UISelectPlatformAndroid],
  ]

  describe.each(platform)('platform %s', (_, Component) => {
    beforeEach(() => {
      mockUISelectPlatform.mockImplementation((props) => (
        <Component {...props} />
      ))
    })

    it('renders correctly without description', () => {
      const element = render(
        <UIOption
          items={items}
          label="Some text here"
          selectedValue={selectedValue}
        />,
      )

      expect(element.toJSON()).toMatchSnapshot()
    })

    it('renders correctly with description', () => {
      const element = render(
        <UIOption
          description="Some more text here"
          items={items}
          label="Some text here"
          selectedValue={selectedValue}
        />,
      )

      expect(element.toJSON()).toMatchSnapshot()
    })
  })
})
