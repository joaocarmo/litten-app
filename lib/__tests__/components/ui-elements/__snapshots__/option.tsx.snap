// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test for the "UISwitch" component platform android renders correctly with description 1`] = `
[
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "borderBottomColor": "#c4c4c4",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 12,
        "paddingBottom": 12,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 12,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        {
          "color": "#000000",
          "flex": 1,
          "fontSize": 15.75,
          "fontWeight": "400",
        }
      }
    >
      Some text here
    </Text>
    <Text
      style={
        [
          {
            "color": "#000000",
            "fontSize": 14,
            "fontWeight": "200",
          },
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      One
    </Text>
  </View>,
  <Text
    style={
      [
        [
          {
            "color": "#000000",
            "fontSize": 14,
            "fontWeight": "200",
          },
          {
            "marginTop": 8,
          },
        ],
        undefined,
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    Some more text here
  </Text>,
  <RNCAndroidDialogPicker
    items={
      [
        {
          "color": null,
          "enabled": true,
          "label": "One",
          "style": {
            "backgroundColor": null,
            "color": null,
          },
        },
        {
          "color": null,
          "enabled": true,
          "label": "Two",
          "style": {
            "backgroundColor": null,
            "color": null,
          },
        },
      ]
    }
    onSelect={[Function]}
    selected={0}
    style={
      {
        "height": 2000,
        "position": "absolute",
        "top": 0,
        "width": 2000,
      }
    }
  />,
]
`;

exports[`Snapshot test for the "UISwitch" component platform android renders correctly without description 1`] = `
[
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "borderBottomColor": "#c4c4c4",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 12,
        "paddingBottom": 12,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 12,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        {
          "color": "#000000",
          "flex": 1,
          "fontSize": 15.75,
          "fontWeight": "400",
        }
      }
    >
      Some text here
    </Text>
    <Text
      style={
        [
          {
            "color": "#000000",
            "fontSize": 14,
            "fontWeight": "200",
          },
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      One
    </Text>
  </View>,
  <RNCAndroidDialogPicker
    items={
      [
        {
          "color": null,
          "enabled": true,
          "label": "One",
          "style": {
            "backgroundColor": null,
            "color": null,
          },
        },
        {
          "color": null,
          "enabled": true,
          "label": "Two",
          "style": {
            "backgroundColor": null,
            "color": null,
          },
        },
      ]
    }
    onSelect={[Function]}
    selected={0}
    style={
      {
        "height": 2000,
        "position": "absolute",
        "top": 0,
        "width": 2000,
      }
    }
  />,
]
`;

exports[`Snapshot test for the "UISwitch" component platform ios renders correctly with description 1`] = `
[
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "borderBottomColor": "#c4c4c4",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 12,
        "paddingBottom": 12,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 12,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        {
          "color": "#000000",
          "flex": 1,
          "fontSize": 15.75,
          "fontWeight": "400",
        }
      }
    >
      Some text here
    </Text>
    <Text
      style={
        [
          {
            "color": "#000000",
            "fontSize": 14,
            "fontWeight": "200",
          },
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      One
    </Text>
  </View>,
  <Text
    style={
      [
        [
          {
            "color": "#000000",
            "fontSize": 14,
            "fontWeight": "200",
          },
          {
            "marginTop": 8,
          },
        ],
        undefined,
        undefined,
        undefined,
        undefined,
      ]
    }
  >
    Some more text here
  </Text>,
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    supportedOrientations={
      [
        "portrait",
      ]
    }
    transparent={true}
    visible={false}
  />,
]
`;

exports[`Snapshot test for the "UISwitch" component platform ios renders correctly without description 1`] = `
[
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "borderBottomColor": "#c4c4c4",
        "borderBottomWidth": 0.5,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 12,
        "paddingBottom": 12,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 12,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        {
          "color": "#000000",
          "flex": 1,
          "fontSize": 15.75,
          "fontWeight": "400",
        }
      }
    >
      Some text here
    </Text>
    <Text
      style={
        [
          {
            "color": "#000000",
            "fontSize": 14,
            "fontWeight": "200",
          },
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
    >
      One
    </Text>
  </View>,
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    supportedOrientations={
      [
        "portrait",
      ]
    }
    transparent={true}
    visible={false}
  />,
]
`;
